### ISSUE TITLE
Entity context switching

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/wordpoints#479]
This was discussed in https://github.com/WordPoints/hooks-api/issues/63, but punted. The plan was always to do it eventually though, and I think that we're going to find uses for it sooner rather than later. So let's go ahead and get it in early.

**Comment by JDGrimes** [comment]
## Cross-context relationships

Information about how we'd use this to support cross-context entity relationships: https://github.com/WordPoints/hooks-api/issues/63#issuecomment-169154612

Also [in regard to cross-entity relationships, we said](https://github.com/WordPoints/hooks-api/issues/63#issuecomment-170260549):

>We also need to switch contexts so that the context of the primary entity is the current context when we are pulling the related entities from the database.

But really, the broader takeaway is that relationships will need to define how they relate to context. By default we expect the related entities will come form the same context, but if they don't the relationship will have to declare that. Or rather, I suppose that the logic for retrieving the related entity is contained in the relationship class anyway, so all of that would be handled internally, though we might still need to declare it in some way for other APIs, I guess.

Anyway, that is really a separate topic that we'll pick up in a separate issue.

**Comment by JDGrimes** [comment]
## Cross-context entity retrieval

So something to consider is that perhaps we also want a cross-context entity retrieval API. This may usually wrap the context switching API (switch to context, retrieve entity, switch back), but wouldn't necessarily have to in every case. As switching contexts can sometimes be expensive, if it is possible to retrieve entities without doing so, it would be nice to not have that logic hard-coded (or rather, to allow each entity type to override it as necessary/useful). This is related to [the idea of stateless contexts](https://github.com/WordPoints/hooks-api/issues/63#issuecomment-169384326), but is really different, as even entities related to stateful contexts may be able to be retrieved from out of context.

**Comment by JDGrimes** [comment]
The contexts currently extend an abstract class `WordPoints_Entity_Context`, which provides the common interface for them (there isn't a `WordPoints_Entity_ContextI` interface). So we could conceivably just add the context switching methods to that. However, they couldn't be abstract, for backward-compatibility reasons, and yet we also couldn't implement them generically. If we did implement them, we'd have to just return `false`. So I have been thinking that maybe we should instead introduce a `WordPoints_Entity_Context_SwitchableI` interface (or something of the sort), that switchable contexts can implement. That way if the switching method needs to return false to indicate that switching failed, that won't be ambiguous. Or, that is, we will be able to tell whether switching was implemented for that context at all or not. Although I suppose that most all contexts should have switching implemented. But still, interfaces are cleaner.

**Comment by JDGrimes** [comment]
On second thought though, in this case interfaces might just be an over-complication. If we were using a general context interface to start with, we'd have no choice. But currently we aren't. (We could change that, but it should be done in a separate ticket. I guess that we could then even skip the separate switchable interface altogether, and just add those methods on the main interface.)

Anyway, interfaces or not, the bigger question is whether switching should be able to be separate form the contexts, that is, whether there can really be non-switchable contexts, or contexts for which we can't implement a switcher, as opposed to just contexts for which switching hasn't been implemented (yet). If all contexts are switchable, then since we have the opportunity to make the feature universal (even if crippled when not implemented), then it makes sense to do so, I think.

So the question we need to answer is, are all contexts switchable?

**Comment by JDGrimes** [comment]
I guess is really kind of the same question as "are all contexts stateful?" That is, can all contexts be connected to the current state in some way?

And I guess in a way, the answer to that is a blunt "yes". Because even if a context could theoretically not support state-switching by default, we could always implement it ourselves (though that would effect how we retrieve the entities that relate to that context). In other words, we could keep track of our own "current context ID" for that context type within our own internal application state, using globals or static vars, or whatever.

**Comment by JDGrimes** [comment]
And on further thought, even though [we discussed stateless contexts](https://github.com/WordPoints/hooks-api/issues/63#issuecomment-169384326) at length previously, I'm not sure that the idea really makes any sense. Sure, context and state are different, but if a context isn't connected to the state, then is it really a context, and not a relationship? Take term taxonomies as an example. Terms are tied to a particular taxonomy, but is that a relationship or a context? There is no "current taxonomy", so this is stateless. But previously in WordPress, you couldn't retrieve the term without knowing both the taxonomy and the ID. We had thought we could only work around this by creating custom UIDs for the terms by combining the ID and taxonomy. But upon consideration, perhaps this would have been a possible case for a stateless context. On the other hand though, it is different, because the same term could relate to multiple taxonomies. And true contexts imply that things that exist only within them cannot also exist in other contexts. I guess that is "by definition."

**Comment by JDGrimes** [comment]
Anyway, the broader point is that stateless contexts look suspiciously like relationships. There is a catch though. An entity can be retrieved without regard to its relationships, but in cases like this (term taxonomies as an example again), the ID of the "related entity" (or "context) also has to be known. (This wasn't necessarily true for the terms, because we could in theory just ignore the taxonomy entirely and pull up the term regardless, but it would have been hacky. Just set that aside for the moment though.)

But then, what is the difference between a "multi-part ID" (which is another way we could look at this), and a context? Basically, it is the connection to state. Now, as we said above, we could connect the context to the state internally, but this does have its own peculiarities when you start to think about it further. It mean that the current context would always have to be manually set, otherwise there would be no current context at all. Though that really isn't that strange, as it is possible for other contexts to not be set. Like site context isn't set when in context of the network generally. But the point is that non-stateful contexts, even with internally tracked statefulness, would always be out of context by default. So any hook args for entities relating to that context, for example, would have to set the context before constructing the entity object (which means that it would also have to know what it was). However, I don't think that that is egregious, and it demonstrates to me sufficiently that for all such contexts it would be possible to implement this.

In other words, even contexts that weren't naturally stateful could be made switchable.

**Comment by JDGrimes** [comment]
Note also that such a state of affairs, which is essentially, as we said above, having a multi-part ID for an entity, is very unlikely. It isn't a common pattern, and so really shouldn't be of much concern to us, either now or in the future.

**Comment by JDGrimes** [comment]
So I see no reason not to go ahead and use abstract methods (or, not really abstract, but methods on the main class), since we have that opportunity. In other words, we are saying that all contexts must be switchable here, and so they should implement switching functionality. We won't break back-compat with those that don't (in this release), but going forward they are all expected to. (And we'll likely also introduce an interface in a separate ticket, and plan to switch to it in 3.0.0).

**Comment by JDGrimes** [comment]
## Method naming

When it comes to naming conventions, [WP#38098](https://core.trac.wordpress.org/ticket/38098 "Use common naming for context switching functions / classes") is related.

Importantly, we can't use `switch()`, because it is a reserved word in PHP. `switch_to()` is the likely alternative.

The real decision comes when we decide what to call the function the "deswitches", or shifts back to the previous state. In our case, I think that `switch_back()` might be a good candidate.

**Comment by JDGrimes** [comment]
## Support switching back?

I guess that really though, we could ask ourselves whether we should even have such a method, or whether switching back to the prior state is a universal feature of contexts. We could in theory just introduce the `switch_to()` method, and when people wanted to switch back, they would have to call `switch_to()` again with the previous context ID. But I guess even if that isn't a universal feature implemented for every context, it is a universal use-case. So we should just implement it internally as necessary, and make it a universal feature of our context API.

**Comment by JDGrimes** [comment]
## Switching out of context?

Interesting question, should it be possible to switch to out-of-context? For some contexts that isn't really possible though. It really only applies to stateless contexts. For other contexts we can't switch out in that way. But for them, I suppose that since we will be implementing the context switching internally anyway (in many cases) we can just implement the option to switch back to the stateless state as well. So we don't really need to take it into account for our average context switching code.

**Comment by JDGrimes** [comment]
## Network switching

Hmmm, something that I hadn't considered before, is that by default there is no way to switch networks in WordPress. So context switching for networks actually can't be implemented by default. I'm not sure if even multi-network plugin support network switching. So, I'm reconsidering whether it makes sense to say "all contexts are switchable", when this one doesn't appear to be. Which means that maybe it would indeed by cleaner and better to go with the interface instead. Then we'd register a custom network context object that implemented the switchable interface when switching was available. Though we can essentially accomplish the same thing now, without having to register a custom object, by checking for the existence of network-switching functions (if there are any such things) in the method of our class. Which is really simpler.

**Comment by JDGrimes** [comment]
[WP Multi Network](http://wordpress.org/extend/plugins/wp-multi-network/) does indeed implement a `switch_to_network()` function. [WP#25293](https://core.trac.wordpress.org/ticket/25293) also contemplates introducing such a function to WordPress. However, it doesn't exist yet, of course.

**Comment by JDGrimes** [comment]
Another thing that we hadn't considered before, is that when WordPress isn't installed as multisite, it isn't possible to switch to a different site or network either. Though I guess in a way that is a different problem, since no other sites or networks could exist then. It isn't that there isn't any switching code for that implemented, it is that there is nothing to switch to. However, with network switching, it is possible that there would be networks that existed, but we wouldn't be able to switch to them.

**Comment by JDGrimes** [comment]
## Failure causes

I think I'm still OK with adding these methods to the abstract class instead of putting them on the interface, the only question would be whether we feel that there is a need to distinguish between different reasons for an inability to switch contexts. Do we need to be able to differentiate between "context doesn't exist" and "switching isn't implemented for this context type".

I guess it really comes down to the fact that an interface allows us to make the check whether switching is possible sooner, without having to actually attempt it in order to find that out. Which makes for cleaner logic, since it allows us to completely disable a feature that requires switching, for example, rather than juts finding out mid-way through that it is crippled.

And generally speaking, I don't think that switching failures for other reasons will be a common thing. They would usually indicate "context ID invalid", rather than "something went wrong" or "switching from the current state is blocked for some reason," or whatever. Though if we really needed that kind of information we could modify the API to provide it in the future. But at the moment I think it is overkill, since we don't even know whether the context switching implementations that we'll be wrapping will allow us to get that kind of fine-grained understanding of a failure in any case anyway. So for now at least it makes sense for this API to be blunt, since it is just a common interface for interacting with contexts in the most general ways that they can be interacted with across the board.

**Comment by JDGrimes** [comment]
I guess that really we should take a look at what kind of return value `switch_to_blog()` and `switch_to_network()` (such as it is at present) have, as that will dictate what kind of return values we can provide from our wrapper methods.

`switch_to_blog()` "always returns true" (as indicated in the inline docs). It can't really fail, and I'm not sure that it ever could or likely ever will be able to in the future either.

`switch_to_network()`, as implemented in WP Multi Network, does the same by default. However, it does offer a second parameter, `$validate`, which will cause `false` to be returned if the network doesn't exist. As there is no patch on the relevant ticket ([WP#29411](https://core.trac.wordpress.org/ticket/29411 "Add multi-network helper functionality")), we can't be sure what this will look like once it goes into core, but WP Multi Network does appear to be the the de facto standard.

**Comment by JDGrimes** [comment]
Incidentally, I just realized that even on multi-network the site IDs are unique to the entire install, not just per-network. This is manifest in the fact that the `blog_id` column of the `blogs` table is a primary (and thus unique) key, even though it exists alongside the `site_id` column (for network IDs). So it turns out that a site can be uniquely identified by its ID, the network ID isn't needed. (At least that is currently the case, but I'm not 100% sure of course about how some multi-network installs might be set up now or in the future.) This basically means that we might not have needed to make the site context a child of the network context. On second thought though, I suppose that it is possible to be in context of a network itself, and not the site as a whole. So the network context would still need to exist. But perhaps it should have been a sibling of the site context, rather than a parent. Anyway, as long as sites can only belong to a single network (and that is certainly the current design, and would take quite a bit to change), we're fine leaving things as they are.

**Comment by JDGrimes** [comment]
So let's keep chugging on without the interface, and see where it leads.

The methods should return a boolean value based on whether the requested context was made the current context successfully or not.

**Comment by JDGrimes** [comment]
## Interface revisited

> I guess it really comes down to the fact that an interface allows us to make the check whether switching is possible sooner, without having to actually attempt it in order to find that out. Which makes for cleaner logic, since it allows us to completely disable a feature that requires switching, for example, rather than juts finding out mid-way through that it is crippled.

I was almost going to say let's do the interface instead after reading this again. However, the only reason to do that would be if we would be likely to ever have logic that would trigger us to switch a context, when switching wouldn't be available. But since switching will likely only ever be triggered by us being presented with an ID that doesn't match the ID of the current context, and since that wouldn't usually ever happen when switching wasn't available, it really isn't likely that we'll ever be in a scenario where switching will fail because no implementation is available. With network switching, for example, we would never be presented with a network ID other than 1 when outside of multi-network installs and network switching wasn't available.

That is assuming, however, that the switching implementation will always accompany the multi-context scenario. And that isn't guaranteed to be the case, as the existence of various multi-network implementations demonstrates. So I think that there may still be an argument for using the interface there. If we are going to have times when we'll want to switch but can't due to a lack of implementation, then the interface is the way that we should go, so that it is possible for us to check whether switching is available at all for a context.

**Comment by JDGrimes** [comment]
Of course, we could simply suggest that switching should be implemented internally in such cases, if it cannot be wrapped. That would tend to be more fragile, however. And yet, again, if there exists multiple contexts, it seems almost certain that switching would go with that, without it multiple contexts wouldn't be usable. Or would they? I suppose if each was entirely self-contained, switching might be seen as unnecessary. The only reason that we'd come up against a need for it is because we have things like points logs that span across contexts. But then again, it does seem like in such a scenario, it would indeed be worth it for us to implement switching ourselves, because otherwise things like viewing points logs out-of-context would break.

**Comment by JDGrimes** [comment]
My reason for wanting to avoid the interface is that it will make the API more complex. We then have to always check whether a context can be switched before switching, which will be a pain (especially considering that that will be a primary use of the context objects going forward). 

I was going to say that we could always switch to an interface later, but it really wouldn't be as easy to transition as it would be to introduce it now, since every context would then implement it, even those without switching implemented. Although on the other hand that isn't an issue if most of them will have an implementation. And I suppose that all of them likely will.

**Comment by JDGrimes** [comment]
And yet I'm not intending to implement the network switching ourselves, mind you. Just piggy-back on `switch_to_network()` if it is available, and return false if it is not. I think that it is safe for us to assume that `switch_to_network()`, when implemented, will always have a signature comparable to `switch_to_blog()`. If we find otherwise, we can cry about it later.

**Comment by JDGrimes** [comment]
## Switch back function return values

For the record, both `switch_to_blog()` and `switch_to_network()` (as implemented in WP Multi Network—by the way, the latest version of that plugin as of this writing is 1.8.1) do return false if the stack is empty and there is nothing to switch back to. Otherwise they return true.

**Comment by JDGrimes** [comment]
We also have to decide what to do when not on multisite and `switch_to_blog()` and `restore_current_blog()` are not available. I suppose we'll just return false if `! is_multisite()`.

**Comment by JDGrimes** [comment]
Although, if we do that, we'll return false even when switching to the id of 1, which is the main site. Which means that our logic that uses the API will have to check that it doesn't switch when it is already in the correct context somewhere along the way. Otherwise on a single site install switching will always fail. I suppose that if we wanted to we could make it so that when the switching implementation isn't available for either of these contexts (site or network, the only two that we have at the moment), they only return false if trying to switch to an ID other than 1, but if switching to the ID 1 return true. However, I don't know that that is necessary, since in most cases it does make sense not to switch to a new context if we don't have to, and so switching should just be skipped altogether on single sites, since everything will be in the same context. We could still add such logic to the contexts though, just as a buffer (and note that we wouldn't need to worry about having a stack there, since there would only be one valid context ID).

**Comment by JDGrimes** [comment]
I think for the sake of encouraging better switching practices, it is safe for us not to add special handling for the ID 1 on single sites. This isn't the kind of issue that would be likely to be missed in somebodies code, and it would alert them to the fact that they are switching when they don't need to. While some implementations (`switch_to_blog()` for example) will optimize that anyway, it does still eat up memory by adding to the stack. The function call (and in this case, the call the wrapper method as well), can be avoided entirely by just keeping track of the current site.

At any rate, if we want to add that kind of logic in the future for some reason, I think we could.

**Comment by JDGrimes** [comment]
Note that another option in the future, if we need to, is to add a `can_switch()` method.

**Comment by JDGrimes** [comment]
Now that we've figured out the basic API, there are lots of different ways that we can use it:

- [Cross-context entity relationships](https://github.com/WordPoints/wordpoints/issues/479#issuecomment-251773035).
- [Cross-context entity retrieval API](https://github.com/WordPoints/wordpoints/issues/479#issuecomment-251773185).
- [Support for setting entity contexts](https://github.com/WordPoints/hooks-api/issues/63#issuecomment-169154612) (possibly related to the above).
- Multi-context entity arrays.

Probably other things as well. These should be pursued in separate tickets.

**Comment by JDGrimes** [comment]
I guess at some point we might also want an `is_switched()` method here. I'm not sure how useful that really is, however. It seems like it would be prone to misuse. You can't check whether you need to `switch_back()` just based on whether the context is switched, because the context might have already been switched to start with. So I'm OK with not including this in the API for now. It might be worth investigating how WordPress uses `ms_is_switched()` though, just to be sure that that's the right decision, and there aren't some obvious use-cases that we're missing.

**Comment by JDGrimes** [comment]
How `ms_is_switched()` is used:

- In `update_home_siteurl()`, to delete the rewrite rules from the DB instead of calling `flush_rewrite_rules()`, I assume because that function can affect the filesystem, which isn't switched by `switch_to_blog()`.
- In `_wp_upload_dir()`, again, filesystem related.
- In `get_custom_logo()`, *incorrectly*, to check if it has switched to a different blog internally within the functions. Oops.
- In `get_site_icon_url()`, ditto.
- In `has_custom_logo()`, ditto.

I guess I should report those last three as bugs.

Anyway, I think this proves my point, although it does alert us to something that we should keep in mind: context switching is a DB-related thing, and doesn't necessarily affect other externals, like the filesystem.

**Comment by JDGrimes** [comment]
Working on #535 now, I am realizing that the complexity of nested contexts really necessitates a canonical switching function. Actually, I think that a switching class would allow us to keep track of an internal stack, so that we could skip switching at all when it wasn't necessary. So let's introduce something like that.

**Comment by JDGrimes** [comment]
This brings up an interesting conundrum, and that is how to handle switching back when we have nested contexts. Do we switch the highest one back first, or the lowest one?

When we are switching *to* a context, we switch the highest one first, as that is necessary before we can switch to a lower one. And I guess that switching back really is just switching back *to* a context, and so should follow that same rule. But for a minute I was wondering whether we should switch the lowest one back first, to perform the switching operation in reverse. That is how it was currently being done in my code, but I was second-guessing myself when I went to correct it.

**Comment by JDGrimes** [comment]
Another issue:

When we are switching to a new context, we try to save ourselves from doing extra work by only actually performing the switch if we really need to, that is, only if we given context ID doesn't match the current context ID. This is all well and good when there is only a single context, but with nested contexts it becomes more complicated. What happens when we do need to change a parent context, but then the child context IDs match up? Does that mean that the child context we were asked to switch to was the default automatically instated when switching the parent context? Or does is it possible that the child context is stuck in the old "ghost context" from the previous parent, and the IDs just happen to collide? I suppose that really ghost child contexts could be considered a bug in the switching implementation, but if so, we need to be emphatic about that right from the start.

What caught me off guard though was the thought that this might affect our site context, and cause issues. That is, I was concerned that we'd end up calling on site switching when not in multisite context, if we forced all children to switch, even if the IDs matched, when the parents had switched. Thinking that switching of the site context when not on multisite would cause issues. However, even if we forced the switching of child contexts when parent context had switched (and I'm not not sure we need to), that wouldn't affect this case, since the network context would never get switched when not in multisite context (duh).

**Comment by JDGrimes** [comment]
> I suppose that really ghost child contexts could be considered a bug in the switching implementation, but if so, we need to be emphatic about that right from the start.

I don't think that that is really a defensible position. ~~Contexts don't necessarily have any relation to each other~~ not so, then they wouldn't be children of the parent, but siblings. A child context that truly is a child of the parent, and not a sibling, would automatically be switched by the parent changing. For example, when switching from one site to another, a sub-context of ID 3 would have the same relation to the parent context in each of those sites, respectively. (Remember, contexts are a database/storage related thing.) In other words, whether the IDs "just happen to collide" is irrelevant. The IDs don't collide at all, they actually compliment one another (identical cousins!).

The only issue would be ghost elements that were pulled into the current state that reflect the context of that ID that is a child of the old parent. Think cached info, etc. (though I guess if using WordPress's built-in object cache, that would be switched automatically, but you get the idea). However, if such references aren't corrected on switching of the parent, then calling the switch implementation for that child context aren't likely to have any affect either, since the ID we are asking to switch to will match the current ID, and calling the switcher will (or at least probably *should*) have no effect. Such an arrangement truly would be a bug, and would experience issues any time the parent was switched, it isn't a problem inherent in our own context switching API, but in the API that provided the context to start with.

So let's bee emphatic: *child contexts must properly respond when parent contexts are switched and restored.*

**Comment by JDGrimes** [comment]
Here's an interesting idea: what if we added the context switching code to the contexts app itself? I was thinking that maybe it would give us more direct access to the context objects, as an optimization, but we don't currently use a persistent registry for the contexts. We could change that, but I'm not sure it is necessary. Still, we could then bypass `is_registered()` if we wanted to and just use `isset()` directly, though that doesn't add any benefit if we're still going to have to call `get()` and `instanceof`.

Optimizations aside, the prospect of a single global switcher might have merit, rather than having to construct a new switcher object each time we need to do some switching. However, I can't really think of an argument for either case which doesn't have more to do with optimization and not how the switcher should operate.

We can switch to a global implementation in the future (or vice versa, if we chose to do it global first), but for now let's just stick with it this way and see how things go.

**Comment by JDGrimes** [comment]
I guess we really should pose the question as to whether we would ever need to use multiple switchers at once. Since the state is global, and switching with two separate switchers will affect the global state identically, there seems to me no reason to ever have more than one of them constructed at a time. Indeed, the ability to construct multiple switchers and use them simultaneously might actually lead to their misuse by developers supposing that switching on switcher A and then switcher B, and switching A back before B would have the desired affect; it wouldn't, if A and B reference the any of the same contexts A won't switch back to its prior state while leaving B unswitched. Instead, A would switch back from B's switched state to the A switched state, but it might fail to do this properly due to internal optimizations where we don't call `switch_back()` for contexts that we skipped switching in the first place. In other words, using multiple switchers like this will only create a mess. So having just one global instance gives us the opportunity to avoid such things, while at the same time possibly perform additional optimizations.

**Comment by JDGrimes** [comment]
One thing I'm not sure of here is whether we should break out of our app's `switch_back()` method and return false if it fails to switch back to any of the contexts. At the moment I'm just not check the return value of the contexts' `switch_back()` methods at all. But I don't know whether continuing to switch the sub-contexts back after a parent context has failed to switch back makes sense. I suppose that it can't do much harm either way. And really, I'm not sure how useful it really is for us to pass that information along, since there isn't really anything that the calling code could do about such a failure. Well, I guess maybe that's not entirely true, but they wouldn't know where to start, or that is, they wouldn't know which context failed. SO maybe we should actually abort the back-switching if a context fails to switch back, and re-switch any parent contexts. So then a failure would mean that we didn't modify the context at all, just as a `switch_to()` failure does, rather than leaving us in an intermediate state.

All things considered though, it isn't a big deal either way since failing to switch back is a very rare error that generally would just indicate that the context was already switched back at some point. And if we assume that that's the only cause of such a failure (at least for now—we could always handle this differently in the future), then I suppose it makes sense to ignore failures and go right on, as we won't end up in an intermediate state at all, but rather exactly where we want to be anyway. Well, that isn't guaranteed I suppose, since there may have been several switches on the stack for that context, meaning our target might be switch 2 in the stack instead of an empty stack. As such, I guess we have to admit that we can't really be sure that we aren't in an intermediate state when switching fails. However, in that case, restoring the parent contexts that we've already switched will not be guaranteed to pull us out of the intermediate state either. The switching error still exists, it doesn't go away, and the calling code would still be in a different state than it wants to be. And there isn't really much that it could do about it.

So, rather than adding the complexity of the un-back-switching during a failed switch-back, which would require us to keep more info (the context IDs) than we do currently anyway, let's just go ahead and abort completely when we hit a failure, and return false. No matter what we do, there will probably be nothing that the calling code can do to handle the failed switch-back, rather than explode early. Generally speaking, the calling code probably won't bother to check for a failed switch-back anyway, since it is such a rare event and as we've mentioned, isn't really recoverable anyway if something goes horribly wrong and it does happen. But I guess we should do the caller the courtesy of passing that info along if they are wanting to know about it for some reason.

**Comment by JDGrimes** [comment]
## Global `switch_to()` optimization

I would like to add an optimization to `switch_to()` that would keep a cache of the current IDs of the context based on the last switch, so that it wouldn't have to retrieve them again if used repeatedly. However, we can't assume that no contexts have been switched via some other method in the mean time. So I suppose that kind of optimization should occur in the code that is doing the switching if it can know that it is safe. And that way, it can avoid the overhead of the method call entirely.

I have also considered adding a flag that would allow this to be an internal feature of the switcher that was disabled by default. But again, the overhead of the method call would mean it really would not be as useful as a higher-level optimization, so I don't think that it is worth it.

**Comment by JDGrimes** [comment]
One of the tests for this is currently failing:

```
1) WordPoints_Entity_Context_Test::test_switch_to
Failed asserting that true is false.

/home/travis/build/WordPoints/wordpoints/tests/phpunit/tests/classes/entity/context.php:56
```

**Comment by JDGrimes** [comment]
The problem is that we were expecting the logic of `WordPoints_Entity_Context::switch_to()` to be tested in that test, but we ended up overriding that method in the mock later. Which meant that the mock method was ignoring the parent's actually implementation completely, which is what is supposed to be under test here.
